### PR TITLE
Fix package restore cycle detection

### LINKED ISSUE #9050: NuGet restore cycle detection doesn't work
While investigating https://github.com/dotnet/sdk/pull/31798 I noticed that the NuGet cycle detection added in https://github.com/dotnet/project-system/pull/8239 wasn't detecting a very simple ABABAB cycle.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #9050

In #8239 we added a feature that aimed to detect NuGet package restore loops, and to stop them after a certain number of iterations to prevent endless resource usage.

It turns out that feature never worked due to using a `byte[]` as a key in a dictionary.

This PR:

- Fixes the `byte[]` hashing issue by wrapping the `byte[]` with a new `Hash` type that adds `Equals` and `GetHashCode`, makes the type immutable and provides a little more type safety.
- Pulls all cycle detection code out into a single component, simplifying the `PackageRestoreDataSource` class somewhat.
- Bumps the compiler version we use so that we can start dogfooding C# 12 bits, and uses a primary constructor in two places.
- Changes the message displayed to the user to be hopefully clearer and specify exactly which project is causing the restore failure, which should make further investigation easier.
- Adds some unit tests for the restore detection. Writing these, I'm not sure the detection algorithm we use is the best fit. We will monitor telemetry and listen for feedback on false positives/negatives and iterate as needed.
- Updates namespaces across the code from an incomplete move of code from the VS layer to the host-agnostic layer in #8850, so that everything's consistent. Moves some unit test code in the same direction too.
- Removes some redundant using directives in the language service code that started appearing as warnings.

Before
image

After
image

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/9051)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    INonModalNotificationService userNotificationService)
+    : IPackageRestoreCycleDetector
+{
+    // This algorithm is to detect the pattern A -> B -> A -> B -> A in the most recent N values.
+    // To keep track of the most N recent values we are using a queue of fixed size, where:
+    //   The most recent value is inserted at the front in O(1) time, and the oldest value is removed at the 
+    // back in O(1) time.
+    //   The counter will keep track of how many times values have appeared in the queue consecutively.
+    // i.e.
+    // A -> B -> C -> A -> B -> D -> X -> Y -> X -> Y -> X -> Y -> X
```
Comment (melytc) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs:25]: are we able to detect the cycle earlier? like this:
```
// A -> B -> C -> A -> B -> D -> X -> Y -> X -> Y -> X -> Y -> X
                                |--------------------|
```

**Review by melytc — APPROVED** [review]

**Review by MiYanni — APPROVED** [review]

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/PackageRestore/PackageRestoreCycleDetectorTests.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        Assert.False(await instance.IsCycleDetectedAsync(hash0, CancellationToken.None));
+        Assert.False(await instance.IsCycleDetectedAsync(hash1, CancellationToken.None));
+        Assert.False(await instance.IsCycleDetectedAsync(hash0, CancellationToken.None));
+        Assert.False(await instance.IsCycleDetectedAsync(hash1, CancellationToken.None));
+        Assert.False(await instance.IsCycleDetectedAsync(hash0, CancellationToken.None));
+        Assert.False(await instance.IsCycleDetectedAsync(hash1, CancellationToken.None)); // ignored
+    }
+
+    [Fact]
+    public async Task Cycle_ThreeHashes_Wat()
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/PackageRestore/PackageRestoreCycleDetectorTests.cs:59]: Oops, I meant to give this a better name once I worked out what the code was doing. I'll fix it.

The answer to your question: https://www.destroyallsoftware.com/talks/wat

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs:25]: I can add a test for this. My sense is that this whole comment is factually incorrect.

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/PackageRestore/PackageRestoreCycleDetectorTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/PackageRestore/PackageRestoreCycleDetectorTests.cs:59]: I've rewritten the tests to be easier to author and read, and to have a bit more coverage.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs]: Added a test case and tweaked the comment a little.

**Pull request merged**